File: python/nlx402.py
"""
NLx402 Python SDK

Minimal client for:
  - GET  /api/auth/me
  - GET  /api/metadata
  - GET  /protected
  - POST /verify
  - GET  /protected

Defaults to base URL: https://pay.thrt.ai
"""

from __future__ import annotations

import json
from typing import Any, Dict, Optional, Union

import requests


class Nlx402Error(Exception):
    """
    NLx402 HTTP error wrapper.
    """

    def __init__(self, message: str, status: int, body: Any = None) -> None:
        super().__init__(message)
        self.message = message
        self.status = status
        self.body = body

    def __str__(self) -> str:
        return f"{self.message} (status={self.status}, body={self.body})"


class Nlx402Client:
    """
    Minimal NLx402 client using requests.Session under the hood.
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        base_url: str = "https://pay.thrt.ai",
        session: Optional[requests.Session] = None,
    ) -> None:
        """
        :param api_key: Optional API key for authenticated endpoints.
        :param base_url: Base URL of the facilitator (default: https://pay.thrt.ai).
        :param session: Optional requests.Session to reuse connections.
        """
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.session = session or requests.Session()

    def set_api_key(self, api_key: str) -> None:
        """
        Update / set API key at runtime.
        """
        self.api_key = api_key

    def _request(
        self,
        path: str,
        *,
        method: str = "GET",
        require_api_key: bool = False,
        headers: Optional[Dict[str, str]] = None,
        **kwargs: Any,
    ) -> Any:
        """
        Internal request helper.

        :param path: Path like "/api/auth/me".
        :param method: HTTP method.
        :param require_api_key: Whether x-api-key header is required.
        :param headers: Additional headers.
        :param kwargs: Extra arguments for requests.Session.request().
        :return: Parsed JSON body or raw text.
        :raises Nlx402Error: if response status is not 2xx.
        """
        url = f"{self.base_url}{path}"

        final_headers: Dict[str, str] = dict(headers or {})
        if require_api_key:
            if not self.api_key:
                raise ValueError(
                    "NLx402: API key is required but not set. "
                    "Call set_api_key() or pass it to Nlx402Client(...)."
                )
            final_headers.setdefault("x-api-key", self.api_key)

        response = self.session.request(method, url, headers=final_headers, **kwargs)

        text = response.text
        try:
            body = response.json() if text else None
        except ValueError:
            body = text

        if not response.ok:
            raise Nlx402Error(
                f"NLx402 request failed with status {response.status_code}",
                response.status_code,
                body,
            )

        return body

    def get_auth_me(self) -> Dict[str, Any]:
        """
        GET /api/auth/me

        Introspect current API key.

        Returns:
            {
              "ok": bool,
              "created_at": number,
              "wallet_id": string,
              "selected_mint": string
            }
        """
        return self._request(
            "/api/auth/me",
            method="GET",
            require_api_key=True,
        )

    def get_metadata(self) -> Dict[str, Any]:
        """
        GET /api/metadata

        Public metadata about the facilitator.

        Returns:
            {
              "ok": bool,
              "metadata": {
                "network": string,
                "supported_chains": [string],
                "version": string
              },
              "supported_mints": [string]   # or supported_tokens depending on backend
            }
        """
        return self._request(
            "/api/metadata",
            method="GET",
            require_api_key=False,
        )

    def get_quote(self, total_price: Union[int, float, str, None] = None) -> Dict[str, Any]:
        """
        Step 1 — GET /protected

        Get a quote (amount, mint, nonce, expiry).

        Always sends X-Total-Price, defaulting to 0.5 if not provided.

        :param total_price: Value to send in X-Total-Price header (e.g. 0.002).
                            Defaults to 0.5 if None.
        :return: QuoteResponse dict:
            {
              "amount": string,
              "chain": string,
              "decimals": number,
              "expires_at": number,
              "mint": string,
              "network": string,
              "nonce": string,
              "recipient": string,
              "version": string
            }
        """
        headers: Dict[str, str] = {}

        total = 0.5 if total_price is None else total_price
        headers["x-total-price"] = str(total)

        return self._request(
            "/protected",
            method="GET",
            headers=headers,
            require_api_key=True,
        )

    def verify_quote(self, *, quote: Union[Dict[str, Any], str], nonce: str) -> Dict[str, Any]:
        """
        Step 2 — POST /verify

        Lock/verify a quote. The server re-derives the quote
        server-side and checks that nothing was tampered with.

        :param quote: Object from get_quote() or a JSON string.
        :param nonce: Same nonce from the quote.
        :return: VerifyResponse dict:
            { "ok": bool }
        """
        if not nonce:
            raise ValueError('verify_quote: "nonce" is required')

        if isinstance(quote, str):
            payment_data = quote
        else:
            # Compact JSON for consistency
            payment_data = json.dumps(quote, separators=(",", ":"))

        data = {
            "payment_data": payment_data,
            "nonce": nonce,
        }

        headers = {
            "content-type": "application/x-www-form-urlencoded",
        }

        return self._request(
            "/verify",
            method="POST",
            headers=headers,
            data=data,
            require_api_key=True,
        )

    def get_paid_access(self, *, tx: str, nonce: str) -> Dict[str, Any]:
        if not tx or not nonce:
            raise ValueError('get_paid_access: "tx" and "nonce" are required')

        x_payment = json.dumps({"tx": tx, "nonce": nonce})
        headers = {"x-payment": x_payment}

        return self._request(
            "/protected",
            method="GET",
            headers=headers,
            require_api_key=True,
        )

    def get_and_verify_quote(self, total_price: Union[int, float, str, None] = None) -> Dict[str, Any]:
        """
        Convenience flow: get a quote, immediately verify it.

        You still need to:
          - pass the quote to the client for payment on-chain
          - call get_paid_access(tx=..., nonce=...) after payment

        :param total_price: Optional price for X-Total-Price (defaults to 0.5).
        :return: {
            "quote": QuoteResponse,
            "verify": VerifyResponse
        }
        """
        quote = self.get_quote(total_price=total_price)
        nonce = quote.get("nonce")
        if not nonce:
            raise ValueError("get_and_verify_quote: quote did not contain a 'nonce' field")
        verify = self.verify_quote(quote=quote, nonce=nonce)
        return {"quote": quote, "verify": verify}

File: javascript/nlx402.js
/**
 * NLx402 Node.js SDK
 *
 * Minimal client for:
 *  - GET  /api/auth/me
 *  - GET  /api/metadata
 *  - GET  /protected
 *  - POST /verify
 *  - GET  /protected
 *
 * Defaults to base URL: https://pay.thrt.ai
 */

/**
 * @typedef {Object} Nlx402ClientOptions
 * @property {string} [apiKey]
 * @property {string} [baseUrl]
 * @property {typeof fetch} [fetch]
 */

/**
 * @typedef {Object} AuthMeResponse
 * @property {boolean} ok
 * @property {number} created_at
 * @property {string} wallet_id
 * @property {string} selected_mint
 */

/**
 * @typedef {Object} MetadataResponse
 * @property {boolean} ok
 * @property {{network:string, supported_chains:string[], version:string}} metadata
 * @property {string[]} supported_mints
 */

/**
 * @typedef {Object} QuoteResponse
 * @property {string} amount
 * @property {string} chain
 * @property {number} decimals
 * @property {number} expires_at
 * @property {string} mint
 * @property {string} network
 * @property {string} nonce
 * @property {string} recipient
 * @property {string} version
 */

/**
 * @typedef {Object} VerifyResponse
 * @property {boolean} ok
 */

/**
 * @typedef {Object} PaidAccessResponse
 * @property {boolean} ok
 * @property {{
 *   amount: string,
 *   decimals: number,
 *   mint: string,
 *   nonce: string,
 *   status: string,
 *   tx: string,
 *   version: string
 * }} x402
 */

class Nlx402Error extends Error {
  /**
   * @param {string} message
   * @param {number} status
   * @param {any} [body]
   */
  constructor(message, status, body) {
    super(message);
    this.name = 'Nlx402Error';
    this.status = status;
    this.body = body;
  }
}

class Nlx402Client {
  /**
   * @param {Nlx402ClientOptions} [options]
   */
  constructor(options = {}) {
    this.baseUrl = (options.baseUrl || 'https://pay.thrt.ai').replace(/\/+$/, '');
    this.apiKey = options.apiKey || null;
    this._fetch = options.fetch || globalThis.fetch;

    if (!this._fetch) {
      throw new Error(
        'No fetch implementation found. ' +
          'Provide options.fetch or use Node.js 18+ where global fetch is available.'
      );
    }
  }

  /**
   * Update / set API key at runtime.
   * @param {string} apiKey
   */
  setApiKey(apiKey) {
    this.apiKey = apiKey;
  }

  /**
   * Internal request helper.
   * @param {string} path
   * @param {RequestInit & {requireApiKey?: boolean}} [options]
   * @returns {Promise<any>}
   * @private
   */
  async _request(path, options = {}) {
    const url = `${this.baseUrl}${path}`;
    const { requireApiKey, headers, ...rest } = options;

    /** @type {Record<string,string>} */
    const finalHeaders = {
      ...(headers || {}),
    };

    if (requireApiKey) {
      if (!this.apiKey) {
        throw new Error('NLx402: API key is required but not set. Call setApiKey() or pass in constructor.');
      }
      finalHeaders['x-api-key'] = this.apiKey;
    }

    const res = await this._fetch(url, {
      ...rest,
      headers: finalHeaders,
    });

    let body;
    const text = await res.text();
    try {
      body = text ? JSON.parse(text) : null;
    } catch {
      body = text;
    }

    if (!res.ok) {
      throw new Nlx402Error(
        `NLx402 request failed with status ${res.status}`,
        res.status,
        body
      );
    }

    return body;
  }

  /**
   * GET /api/auth/me
   *
   * Introspect current API key.
   * @returns {Promise<AuthMeResponse>}
   */
  async getAuthMe() {
    return /** @type {Promise<AuthMeResponse>} */ (
      this._request('/api/auth/me', {
        method: 'GET',
        requireApiKey: true,
      })
    );
  }

  /**
   * GET /api/metadata
   *
   * Public metadata about the facilitator.
   * @returns {Promise<MetadataResponse>}
   */
  async getMetadata() {
    return /** @type {Promise<MetadataResponse>} */ (
      this._request('/api/metadata', {
        method: 'GET',
      })
    );
  }

  /**
   * Step 1 — GET /protected
   *
   * Get a quote (amount, mint, nonce, expiry).
   *
   * Always sends X-Total-Price, defaulting to 0.5 if not provided.
   *
   * @param {{ totalPrice?: number|string } } [options]
   *   totalPrice → sends X-Total-Price header, e.g. 0.002
   *
   * @returns {Promise<QuoteResponse>}
   */
  async getQuote(options = {}) {
    /** @type {Record<string,string>} */
    const headers = {};

    // Always send X-Total-Price, default 0.5
    const total = options.totalPrice ?? 0.5;
    headers['x-total-price'] = String(total);

    return /** @type {Promise<QuoteResponse>} */ (
      this._request('/protected', {
        method: 'GET',
        headers,
        requireApiKey: true,
      })
    );
  }

  /**
   * Step 2 — POST /verify
   *
   * Lock/verify a quote. The server re-derives the quote
   * server-side and checks that nothing was tampered with.
   *
   * @param {{ quote: QuoteResponse | string, nonce: string }} params
   *   quote → object from getQuote() or a JSON string
   *   nonce → same nonce from the quote
   *
   * @returns {Promise<VerifyResponse>}
   */
  async verifyQuote(params) {
    const { quote, nonce } = params;
    if (!nonce) {
      throw new Error('verifyQuote: "nonce" is required');
    }

    const paymentData =
      typeof quote === 'string' ? quote : JSON.stringify(quote);

    const body = new URLSearchParams();
    body.set('payment_data', paymentData);
    body.set('nonce', nonce);

    return /** @type {Promise<VerifyResponse>} */ (
      this._request('/verify', {
        method: 'POST',
        body,
        headers: {
          'content-type': 'application/x-www-form-urlencoded',
        },
        requireApiKey: true,
      })
    );
  }

  /**
   * Step 3 — GET /protected with X-Payment
   *
   * Access the protected resource after the user has paid on-chain.
   *
   * @param {{ tx: string, nonce: string }} params
   *   tx    → Solana transaction signature
   *   nonce → quote nonce used when verifying
   *
   * @returns {Promise<PaidAccessResponse>}
   */
  async getPaidAccess(params) {
    const { tx, nonce } = params;
    if (!tx || !nonce) {
      throw new Error('getPaidAccess: "tx" and "nonce" are required');
    }

    const xPayment = JSON.stringify({ tx, nonce });

    return /** @type {Promise<PaidAccessResponse>} */ (
      this._request('/protected', {
        method: 'GET',
        headers: {
          'x-payment': xPayment,
        },
        requireApiKey: true,
      })
    );
  }

  /**
   * Convenience flow: get a quote, immediately verify it.
   *
   * You still need to:
   *  - pass the quote to the client for payment on-chain
   *  - call getPaidAccess({tx, nonce}) after payment
   *
   * @param {{ totalPrice?: number|string }} [options]
   * @returns {Promise<{quote: QuoteResponse, verify: VerifyResponse}>}
   */
  async getAndVerifyQuote(options = {}) {
    const quote = await this.getQuote(options); // will default to 0.5 if not provided
    const verify = await this.verifyQuote({
      quote,
      nonce: quote.nonce,
    });
    return { quote, verify };
  }
}

module.exports = {
  Nlx402Client,
  Nlx402Error,
};
